crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,849 @@
+/// be accessed otherwise.
+unsafe fn mut_ptr_as_ref<'a, T>(ptr: *mut T) -> &'a mut T {
+    check_ptr(ptr).unwrap();
+    let as_mut_ref = unsafe { ptr.as_mut() };
+    as_mut_ref.unwrap() // we know the pointer is not null, hence we can safely unwrap
+}
+
+/// @ingroup QkSparseObservable
+/// Construct the zero observable (without any terms).
+///
+/// @param num_qubits The number of qubits the observable is defined on.
+///
+/// @return A pointer to the created observable.
+///
+/// # Example
+///
+///     QkSparseObservable *zero = qk_obs_zero(100);
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_obs_zero(num_qubits: u32) -> *mut SparseObservable {
+    let obs = SparseObservable::zero(num_qubits);
+    Box::into_raw(Box::new(obs))
+}
+
+/// @ingroup QkSparseObservable
+/// Construct the identity observable.
+///
+/// @param num_qubits The number of qubits the observable is defined on.
+///
+/// @return A pointer to the created observable.
+///
+/// # Example
+///
+///     QkSparseObservable *identity = qk_obs_identity(100);
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_obs_identity(num_qubits: u32) -> *mut SparseObservable {
+    let obs = SparseObservable::identity(num_qubits);
+    Box::into_raw(Box::new(obs))
+}
+
+/// @ingroup QkSparseObservable
+/// Construct a new observable from raw data.
+///
+/// @param num_qubits The number of qubits the observable is defined on.
+/// @param num_terms The number of terms.
+/// @param num_bits The total number of non-identity bit terms.
+/// @param coeffs A pointer to the first element of the coefficients array, which has length
+///     ``num_terms``.
+/// @param bit_terms A pointer to the first element of the bit terms array, which has length
+///     ``num_bits``.
+/// @param indices A pointer to the first element of the indices array, which has length
+///     ``num_bits``. Note that, per term, these *must* be sorted incrementally.
+/// @param boundaries A pointer to the first element of the boundaries array, which has length
+///     ``num_terms + 1``.
+///
+/// @return If the input data was coherent and the construction successful, the result is a pointer
+///     to the observable. Otherwise a null pointer is returned.
+///
+/// # Example
+///
+///     // define the raw data for the 100-qubit observable |01><01|_{0, 1} - |+-><+-|_{98, 99}
+///     uint32_t num_qubits = 100;
+///     uint64_t num_terms = 2;  // we have 2 terms: |01><01|, -1 * |+-><+-|
+///     uint64_t num_bits = 4; // we have 4 non-identity bits: 0, 1, +, -
+///
+///     complex double coeffs[2] = {1, -1};
+///     QkBitTerm bits[4] = {QkBitTerm_Zero, QkBitTerm_One, QkBitTerm_Plus, QkBitTerm_Minus};
+///     uint32_t indices[4] = {0, 1, 98, 99};  // <-- e.g. {1, 0, 99, 98} would be invalid
+///     size_t boundaries[3] = {0, 2, 4};
+///
+///     QkSparseObservable *obs = qk_obs_new(
+///         num_qubits, num_terms, num_bits, coeffs, bits, indices, boundaries
+///     );
```

Comment:
Super minor documentation stuff (no need to change now): it'd be cool to use GitHub-style fenced code blocks, assuming Doxygen supports them (I think it does), if not only because it's easier to see where the code is, and my editor will highlight the internal code for me ;)

Reply:
It seems it does https://www.doxygen.nl/manual/markdown.html ! Also in a fashion that I've never seen code blocks before lol but the triple tick also works 👍🏻